src/syntax_analysis.rs

```diff
@@ -0,0 +1,192 @@
+
+/// Generates independent [SyntaxSet]s after analyzing dependencies between syntaxes
+/// in a [SyntaxSetBuilder], and then prints the reults.
+pub(crate) fn print_syntax_dependencies(syntax_set_builder: &SyntaxSetBuilder) {
+    println!("Constructing independent SyntaxSets...");
+    let independent_syntax_sets = build_independent_syntax_sets(syntax_set_builder);
+
+    println!("Independent SyntaxSets:");
+    for syntax_set in independent_syntax_sets {
+        let names = syntax_set
+            .syntaxes()
+            .iter()
+            .map(|syntax| &syntax.name)
+            .collect::<Vec<&String>>();
+        println!("{:?}", names);
+    }
+}
+
+/// Analyzes dependencies between syntaxes in a [SyntaxSetBuilder].
+/// From that, it builds independent [SyntaxSet]s.
+fn build_independent_syntax_sets(syntax_set_builder: &SyntaxSetBuilder) -> Vec<SyntaxSet> {
+    let syntaxes = syntax_set_builder.syntaxes();
+
+    // Build data structures we need for dependency resolution
+    let (syntax_to_dependencies, dependency_to_syntax) = generate_maps(syntaxes);
+
+    // Create one independent SyntaxSet from each (non-hidden) SyntaxDefinition
+    syntaxes
+        .iter()
+        .filter_map(|syntax| {
+            if syntax.hidden {
+                return None;
+            }
+
+            let mut builder = SyntaxSetDependencyBuilder::new();
+            builder.add_with_dependencies(syntax, &syntax_to_dependencies, &dependency_to_syntax);
+            Some(builder.build())
+        })
+        .collect()
+}
+
+/// In order to analyze dependencies, we need two key pieces of data.
+/// First, when we have a [Dependency], we need to know what [SyntaxDefinition] that
+/// corresponds to. Second, when we have a [SyntaxDefinition], we need to know
+/// what dependencies it has. This functions generates that data for each syntax.
+fn generate_maps(syntaxes: &[SyntaxDefinition]) -> (SyntaxToDependencies, DependencyToSyntax) {
+    let mut syntax_to_dependencies = HashMap::new();
+    let mut dependency_to_syntax = HashMap::new();
+
+    for syntax in syntaxes {
+        let SyntaxDefinition { name, scope, .. } = syntax;
+
+        syntax_to_dependencies.insert(name.clone(), dependencies_for_syntax(syntax));
+
+        dependency_to_syntax.insert(Dependency::ByName(name.clone()), syntax);
+        dependency_to_syntax.insert(Dependency::ByScope(*scope), syntax);
+    }
+
+    (syntax_to_dependencies, dependency_to_syntax)
+}
+
+/// Gets what external dependencies a given [SyntaxDefinition] has.
+/// An external dependency is another `.sublime-syntax` file.
+/// It does that by looking for variants of the following YAML patterns:
+/// - `include: C.sublime-syntax`
+/// - `embed: scope:source.c`
+fn dependencies_for_syntax(syntax: &SyntaxDefinition) -> Vec<Dependency> {
+    let mut dependencies: Vec<Dependency> = syntax
+        .contexts
+        .values()
+        .map(|context| &context.patterns)
+        .flatten()
```

Comment:
You can use [`flat_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) here and below.

Reply:
Good catch. That's a very nice simplification.